webtau-cli/src/main/java/com/twosigma/webtau/cli/Cli.java

```diff
@@ -58,29 +67,75 @@ public void run(String command, CliValidationExitCodeOutputHandler handler) {
 
     public void run(String command, ProcessEnv env, CliValidationExitCodeOutputHandler handler) {
         cliStep(command, env,
-                (runResult) -> handler.handle(exitCode(runResult), cliOutput(runResult), cliError(runResult)));
+                (validationResult) -> handler.handle(
+                        exitCode(validationResult),
+                        validationResult.getOut(),
+                        validationResult.getErr()));
+
+    }
+
+    public CliValidationResult getLastValidationResult() {
+        return lastValidationResult.get();
     }
 
-    private void cliStep(String command, ProcessEnv env, Consumer<ProcessRunResult> validationCode) {
-        TestStep.createAndExecuteStep(null,
+    private void cliStep(String command, ProcessEnv env, Consumer<CliValidationResult> validationCode) {
+        CliValidationResult validationResult = new CliValidationResult(command);
+
+        TestStep<Object, Void> step = TestStep.createStep(null,
                 tokenizedMessage(action("running cli command "), stringValue(command)),
                 () -> tokenizedMessage(action("ran cli command"), stringValue(command)),
-                () -> {
-                    ProcessRunResult runResult = ProcessUtils.run(command, env.getEnv());
-                    if (runResult.getErrorReadingException() != null) {
-                        throw new RuntimeException(runResult.getErrorReadingException());
-                    }
-
-                    if (runResult.getOutputReadingException() != null) {
-                        throw new RuntimeException(runResult.getOutputReadingException());
-                    }
-
-                    validationCode.accept(runResult);
-                });
+                () -> runAndValidate(validationResult, command, env, validationCode));
+
+        try {
+            step.execute(StepReportOptions.REPORT_ALL);
+        } finally {
+            step.addPayload(validationResult);
+            lastValidationResult.set(validationResult);
+        }
+    }
+
+    private void runAndValidate(CliValidationResult validationResult,
```

Comment:
There is, at first glance, a fair bit of overlap between cli and http code in general.  Do you think there is room for consolidation/refactoring or do you think it's sufficiently different that such refactoring would lead to clumsy usage?

Reply:
felt to me like it will be clumsy. But it could be just evening coding.

Reply:
You might be right.  Either way, perhaps best tackled in a separate PR.